third_party/move/tools/move-linter/src/model_ast_lints/aborting_overflow_checks.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use move_compiler_v2::external_checks::ExpChecker;
+use move_model::{
+    ast::{Exp, ExpData, Operation},
+    model::FunctionEnv,
+};
+
+const MSG: &str =
+    "This looks like a C-style overflow check. In Move, overflows automatically abort.";
+
+#[derive(Default)]
+
+pub struct AbortingOverflowChecks;
+
+impl ExpChecker for AbortingOverflowChecks {
+    fn get_name(&self) -> String {
+        "aborting_overflow_checks".to_string()
+    }
+
+    fn visit_expr_pre(&mut self, env: &FunctionEnv, expr: &ExpData) {
+        let g_env = env.env();
+        if is_c_like_check_cmp(&expr) {
+            self.report(g_env, &g_env.get_node_loc(expr.node_id()), MSG);
+        }
+    }
+}
+
+/// Checks if an Exp is a binary comparison and looks if the operands
+/// of the comparison are formatted like a C overflow check:
+/// ```C
+///     if (a > a + b) {
+///         //...
+///     }
+/// ```
+/// It checks for (Gt | Lt) operands, with one side matching (Add | Sub) operations.
+/// Excludes patterns like `a > a - b` and `a - b < a` as they can be true, false or abort.
+fn is_c_like_check_cmp(exp: &ExpData) -> bool {
+    let ExpData::Call(_, operation, operands_vec) = exp else {
+        return false;
+    };
+
+    if !op_is_strict_cmp(operation) || operands_vec.len() != 2 {
+        return false;
+    }
+
+    let (lhs_exp, rhs_exp) = (&operands_vec[0], &operands_vec[1]);
+
+    let lhs_add_sub_operands = get_add_or_sub_operands(lhs_exp);
+    let rhs_add_sub_operands = get_add_or_sub_operands(rhs_exp);
+
+    // Check if one side of the expression contains one expression "equal" to the other side
+    let has_matches = has_matching_operand(lhs_add_sub_operands, rhs_exp)
+        || has_matching_operand(rhs_add_sub_operands, lhs_exp);
+
+    has_matches
+        && !expression_has_different_results(
```

Comment:
nit: seems like this function calls `has_matching_operand` again, and it might be possible to eliminate this by refactoring the code. If after considering the comment, you feel the existing code is better, I am okay with that.

Reply:
I rewrote some expressions to make them more clear, it was weird calling again `has_matching_operand`